### DISCUSSION TITLE
PHP 8.1 issues

### DISCUSSION THREAD

**Discussion by prewebdesign** [wintercms/winter#843] [Q&A]
Hi all - newbie alert! Migrated a client over to Winter 1.2 back end of 2022 and am running PHP 8.0.28 with IONOS on a managed dedicated server. When making the switch to 8.1 through IONOS' control panel, the site goes down. Tried forcing an update but can't see anything that needs changing. Is there something simple I am missing?

**Comment by bennothommo** [comment]
@prewebdesign are there any logs that might give a hint as to what is happening?

**Reply by prewebdesign** [reply]
Initially I thought it was a MySQL issue as I was still running 5.5, but have migrated the DB to 5.7 and updated the database.php file but still no joy when switching to 8.1. Is it best just to start from scratch and copy the theme files over?

**Reply by LukeTowers** [reply]
@prewebdesign ideally you should track down the error log file to see what's going on. 

You should try to use composer to `composer update` after verifying that the output of `php --version` is 8.1. Then try setting config/app.php -> debug to true to see if it'll give you a better error message

**Reply by prewebdesign** [reply]
During inheritance of ArrayAccess: Uncaught ErrorException: Return type of Illuminate\Pagination\AbstractPaginator::offsetExists($key) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Pagination/AbstractPaginator.php:599 Stack trace: #0 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Pagination/Paginator.php(14): Illuminate\Foundation\Bootstrap\HandleExceptions->handleError(8192, 'Return type of ...', '/homepages/19/d...', 599) #1 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/composer/ClassLoader.php(571): include('/homepages/19/d...') #2 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/composer/ClassLoader.php(428): Composer\Autoload\includeFile('/homepages/19/d...') #3 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Pagination/PaginationServiceProvider.php(32): Composer\Autoload\ClassLoader->loadClass('Illuminate\\Pagi...') #4 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(616): Illuminate\Pagination\PaginationServiceProvider->register() #5 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Foundation/ProviderRepository.php(75): Illuminate\Foundation\Application->register(Object(Illuminate\Pagination\PaginationServiceProvider)) #6 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/winter/storm/src/Foundation/Application.php(401): Illuminate\Foundation\ProviderRepository->load(Array) #7 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/winter/storm/src/Foundation/Application.php(409): Winter\Storm\Foundation\Application->registerConfiguredProviders(true) #8 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/RegisterProviders.php(17): Winter\Storm\Foundation\Application->registerConfiguredProviders() #9 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/winter/storm/src/Foundation/Application.php(127): Illuminate\Foundation\Bootstrap\RegisterProviders->bootstrap(Object(Winter\Storm\Foundation\Application)) #10 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(156): Winter\Storm\Foundation\Application->bootstrapWith(Array) #11 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(140): Illuminate\Foundation\Http\Kernel->bootstrap() #12 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request)) #13 /homepages/19/d914619079/htdocs/mepacollege.com/prewebCMS/index.php(43): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request)) #14 {main}

**Reply by prewebdesign** [reply]
I have tried the following when logged in to the website root over SSH (via PuTTY)

/usr/bin/php8.1-cli php artisan winter:update
/usr/bin/php8.1-cli php composer update
/usr/bin/php8.1-cli php artisan composer update

Nothing seems to have worked - still getting the error message. The reason for the /usr/bin/php8.1-cli is because when trying to use command line install originally for Winter/composer with IONOS, this was what they told me I had to use. If I put in php --version in the website root - it appears as PHP 4.4.9.

I use the web installer as I am not 100% knowledgeable with root commands.

**Reply by prewebdesign** [reply]
(uiserver):u*********:~/***********.com/prewebCMS$ php artisan winter:update
Content-type: text/html

<br />
<b>Warning</b>:  Unexpected character in input:  '\' (ASCII=92) state=1 in <b>/homepages/19/d914619079/htdocs/***********.com/prewebCMS/artisan</b> on line <b>31</b><br />
<br />
<b>Parse error</b>:  syntax error, unexpected T_STRING in <b>/homepages/19/d914619079/htdocs/***********.com/prewebCMS/artisan</b> on line <b>31</b><br />

**Reply by LukeTowers** [reply]
yikes, PHP 4.4.9 is still set as your server's default CLI PHP version? That's incredibly concerning and I would look for a new host TBH.

As far as solving your immediate issue goes, please edit the index.php to have the following lines at the top immediately after the first line of `<?php`

```php
echo phpinfo();
die;
```

Then share your results. If it says the webroot is set to PHP 8.1, then try running the following inside of your project root:

```bash
rm -rf vendor/ &&
rm -rf modules/ &&
/usr/bin/php8.1-cli /usr/bin/composer update;
```

**Reply by prewebdesign** [reply]
phpinfo() results...

![image](https://user-images.githubusercontent.com/125877643/220558820-c9d4a984-fd18-4902-9713-c2ce9a342020.png)

**Reply by prewebdesign** [reply]
![image](https://user-images.githubusercontent.com/125877643/220559394-962b2eec-d576-407d-afd6-c69c853ccafc.png)

![image](https://user-images.githubusercontent.com/125877643/220560249-8ddbfcc1-6f1f-4dd0-b29b-999acfee81d9.png)

**Reply by prewebdesign** [reply]
Any update to this please @LukeTowers or @bennothommo - is there any easy way to determine what version of Winter is being used within the Dashboard?

I am currently running PHP8.0 on the majority of my installs and need to update to 8.1 without having to reinstall all CMS' one by one? Does the CMS not auto-update to the latest version through the System > Updates & Plugins?

TIA.

**Reply by LukeTowers** [reply]
@prewebdesign we're still working on backend updating system, we just launched the new website and documentation so the marketplace and backend updating system is the next step that we're working on now.

Do you have composer installed on your server?

**Comment by prewebdesign** [comment]
That’s good – is there a way for me to identify what version I am running?

In terms of composer, I only ever use the web installer as I have always had CLI issues with my web host using SSH so would prefer web interface update if possible.

Thanks

**Reply by LukeTowers** [reply]
If you can get terminal access you can run `artisan winter:version`. If you don't, you can try triggering the code by adding it to a new CMS page and then visiting that page in your browser:
```twig
title="Get Winter Version"
url="/version"
==
<?php
function onStart()
{
    $build = \System\Classes\UpdateManager::instance()->setBuildNumberManually();
    $this['version'] = $build['build'];
}
?>
==
{{ version }}
```

**Comment by prewebdesign** [comment]
Thanks Luke. I think the vast majority of my installs are currently running 1.1.9 on PHP8.0 – without terminal access, how can I update these?

Do all of them have to be manual updates? i.e. rebuild the whole CMS with the latest version to get the built-in update?

**Reply by LukeTowers** [reply]
The easiest way right now would be to use the web installer to create new Winter instances and then copy over your config values (ideally you don't copy the files themselves as there have been some changes), plugins, themes, and storage folders from the old versions to the new versions. 

Our goal once the marketplace is functional is to have an updater still working for the older versions of Winter (and October) so that you can one-click upgrade to the latest Winter and then use the new composer-based updates behind the scenes from then on, but that might get tricky as PHP versions continue to get deprecated.

**Comment by prewebdesign** [comment]
Thanks for letting me know. I have been doing that anyway, but I wondered whether there were plans to introduce the auto-update feature as it is time consuming when running multiple installations! With PHP8.0 community support ending in December this year, I need to get a move on…

**Reply by LukeTowers** [reply]
There definitely is plans to restore the auto-update feature, it'll just take some time :)

**Comment by prewebdesign** [comment]
Thanks Luke – re. updating config values – can I assume there are no database changes? I can continue to use the current DBs?

 

Thanks
Pete

 



 

Peter Ellis • 07590 196914 •  <https://www.prewebdesign.co.uk/> https://www.prewebdesign.co.uk

 

We would love your feedback.  <https://g.page/r/CXzXIBJu3w7ZEAg/review> Leave a review »

**Reply by LukeTowers** [reply]
You should always run `php artisan migrate` after running an update, but if you don't then logging into the backend should trigger it automatically for you.
